waltz-tools/src/main/java/com/wepay/waltz/tools/cluster/ClusterCli.java

```diff
@@ -364,6 +378,103 @@ private void buildStoragePartitionValidationResult(int numPartitions,
             }
         }
 
+        /**
+         * Builds REPLICA_RECOVERY_STATUS {@code PartitionValidationResults} for all the replicas in the
+         * cluster.
+         * The validation result is set to success if and only if:
+         * 1. The set of replicas in replica assignments is the same as the set of replicas in replica states.
+         * 2. For each replica, the session id in the replica states equals the session id in partitionMetadata,
+         * and closingHighWaterMark is UNRESOLVED.
+         *
+         * @param numPartitions Total number of partitions in the cluster.
+         * @param partitionRoot ZNode used to fetch replica states for each partition.
+         * @param replicaAssignments Replica assignments for all the partitions.
+         * @param zkClient ZooKeeper Client used to fetch partition metadata.
+         * @param partitionsValidationResultsList a {@code List<PartitionValidationResults>} that contains all the
+         *                                       validation results.
+         */
+        private void buildStorageRecoveryCompleteValidationResult(int numPartitions,
+                                                                  ZNode partitionRoot,
+                                                                  ReplicaAssignments replicaAssignments,
+                                                                  ZooKeeperClient zkClient,
+                                                                  List<PartitionValidationResults> partitionsValidationResultsList) {
+            // A set that contains all replica ids in replica assignments
+            Set<ReplicaId> assignmentReplicaIdSet = new TreeSet<>();
+            for (Map.Entry<String, int[]> entry : replicaAssignments.replicas.entrySet()) {
+                for (int partitionId : entry.getValue()) {
+                    assignmentReplicaIdSet.add(new ReplicaId(partitionId, entry.getKey()));
```

Comment:
Instead of creating `new ReplicaId's` (not sure if its a right way to do it), I was wondering if it would be better if you can store the replica assignment mapping in a different way. For eg. `Map<Integer,Set<String>>` where key can be the partitionId and value can be replicas of that partition. Or any other different way. 

You can make use of `getZkPartitionToStorageNodeMapping(...)` method as well, and also for simplicity you can change the return type from `Map<Integer, List<String>>` to `Map<Integer, Set<String>>`.